src/projections/sch.cpp

```diff
@@ -48,62 +47,49 @@ struct pj_opaque {
     double transMat[9];
     double xyzoff[3];
     double rcurv;
-    GeocentricInfo sph;
-    GeocentricInfo elp_0;
+    PJ* cart;
+    PJ* cart_sph;
 };
 } // anonymous namespace
 
 PROJ_HEAD(sch, "Spherical Cross-track Height") "\n\tMisc\n\tplat_0= plon_0= phdg_0= [h_0=]";
 
 static PJ_LPZ sch_inverse3d(PJ_XYZ xyz, PJ *P) {
-    PJ_LPZ lpz = {0.0, 0.0, 0.0};
     struct pj_opaque *Q = static_cast<struct pj_opaque*>(P->opaque);
-    double temp[3];
-    /* Local lat,lon using radius */
-    double pxyz[] = {
-        xyz.y * P->a / Q->rcurv,
-        xyz.x * P->a / Q->rcurv,
-        xyz.z
-    };
 
-    if( pj_Convert_Geodetic_To_Geocentric( &(Q->sph), pxyz[0], pxyz[1], pxyz[2], temp, temp+1, temp+2) != 0) {
-            proj_errno_set(P, PJD_ERR_TOLERANCE_CONDITION);
-            return lpz;
-    }
+    PJ_LPZ lpz;
+    lpz.lam = xyz.x * (P->a / Q->rcurv);
+    lpz.phi = xyz.y * (P->a / Q->rcurv);
+    lpz.z = xyz.z;
+    xyz  =  Q->cart_sph->fwd3d (lpz, Q->cart_sph);
+    double temp[] = {xyz.x, xyz.y, xyz.z};
 
     /* Apply rotation */
-    pxyz[0] = Q->transMat[0] * temp[0] + Q->transMat[1] * temp[1] + Q->transMat[2] * temp[2];
-    pxyz[1] = Q->transMat[3] * temp[0] + Q->transMat[4] * temp[1] + Q->transMat[5] * temp[2];
-    pxyz[2] = Q->transMat[6] * temp[0] + Q->transMat[7] * temp[1] + Q->transMat[8] * temp[2];
+    double pxyz[] = {
+        Q->transMat[0] * temp[0] + Q->transMat[1] * temp[1] + Q->transMat[2] * temp[2],
+        Q->transMat[3] * temp[0] + Q->transMat[4] * temp[1] + Q->transMat[5] * temp[2],
+        Q->transMat[6] * temp[0] + Q->transMat[7] * temp[1] + Q->transMat[8] * temp[2]
+    };
```

Comment:
I think it would be better to simply use `xyz` here instead of `temp`. `temp` is a bit misleading as it doesn't really say what it is, except that it is temporary.  I can see why it was simplest to just keep using the temp variable from the old code but this is a good opportunity to make the code a bit easier to read. If you still want to keep the vector-feel of using indices `xyz.v[0]` (etc) can be used.

This of course also applies in the forward function below.

Reply:
yes, I was considering that too, but wanted to make the port to the cart API more obvious. I've added a second commit that simplifies the code